packages/core/src/publicCredential/PublicCredential.chain.ts

```diff
@@ -0,0 +1,230 @@
+  return (
+    extrinsics
+      // Filter out failed extrinsics
+      .filter(({ dispatchError }) => dispatchError === undefined)
+      // Filter out extrinsics that don't contain a `CredentialStored` event with the right credential ID
+      .filter(({ events }) =>
+        events.some(
+          (event) =>
+            api.events.publicCredentials.CredentialStored.is(event) &&
+            event.data[1].toString() === credentialId
+        )
+      )
+      // If there is more than one (e.g., same credential issued multiple times in the same block) it should not matter since the ID is generated over the content, hence same ID -> same content.
+      // Nevertheless, take only the last one, if present, as that is for sure what ended up being in the blockchain state.
+      .pop()?.extrinsic ?? null
+  )
+}
+
+/**
+ * Decodes the public credential details returned by `api.call.publicCredential.getCredential()`.
+ *
+ * This is the **only** secure way for users to retrieve and verify a credential.
+ * Hence, calling `api.call.publicCredentials.getCredential(credentialId)` and then passing the result to this function is the only way to trust that a credential with a given ID is valid.
+ *
+ * @param credentialId Credential ID to use for the query. It is required to complement the information stored on the blockchain in a [[PublicCredentialsCredentialsCredentialEntry]].
+ * @param publicCredentialEntry The raw public credential details from blockchain.
+ * @returns The [[IPublicCredential]] as the result of combining the on-chain information and the information present in the tx history.
+ */
+export async function credentialFromChain(
+  credentialId: HexString,
+  publicCredentialEntry: Option<PublicCredentialsCredentialsCredentialEntry>
+): Promise<IPublicCredential> {
+  const api = ConfigService.get('api')
+
+  const { blockNumber, revoked } = publicCredentialEntry.unwrap()
+
+  const extrinsic = await retrievePublicCredentialCreationExtrinsicsFromBlock(
+    api,
+    credentialId,
+    blockNumber
+  )
+
+  if (extrinsic === null) {
+    throw new SDKErrors.PublicCredentialError(
+      `The block number as specified in the provided credential entry (${blockNumber}) does not have any extrinsic that includes a credential creation.`
+    )
+  }
+
+  if (!api.tx.did.submitDidCall.is(extrinsic)) {
+    throw new SDKErrors.PublicCredentialError(
+      'Extrinsic should be a did.submitDidCall extrinsic'
+    )
+  }
+
+  const [extrinsicCalls, extrinsicDidOrigin] = [
+    flattenCalls(api, extrinsic.args[0].call),
+    didFromChain(extrinsic.args[0].did),
+  ]
+
+  let credentialInput: IPublicCredential | undefined
+
+  // Iterate over the calls in the extrinsic to find the right one, and re-create the issued public credential.
+  // If more than a call are present, it always consider the last one as the valid one.
+  extrinsicCalls.forEach((call) => {
+    if (api.tx.publicCredentials.add.is(call)) {
```

Comment:
I have several suggestions on this loop. First, early returns reduce nesting and improve code clarity. Second, if we’re only interested in the last one, we could reverse the array and look for the first one. Finally, I have a hunch that splitting the loop into several calls to `filter` and `map` might increase clarity and readability even further.

Reply:
Maybe even `find` instead of `filter` given that it's built to return on the first match?

Reply:
https://github.com/KILTprotocol/sdk-js/pull/677/commits/e113cbbc643026f3dadf6e2857bff733dd69d924#diff-3228260bd7a8a1c0fb1e0b94efbc0b692518737b171dda86329b57db740703bbR100

Best suggestion of this PR, nothing against the other ones tho 😄

Reply:
Looks good now!